index.js

```diff
@@ -84,6 +84,192 @@ app.post('/add/:quizName/:count', urlencodedParser, async (req, res) => {
     </html>`)    
 })
 
+////
+////
+////These are the /take routes
+
+app.post('/take', urlencodedParser, async(req, res) => {
+    console.log("This makes sure that this is running properly")  // TEMP
+    
+    var quizName = req.body.userName
+    const inCase = 'Sorry, this quiz does not exist.'
+    if (quizName == '') {
+        quizName = inCase
+    }
+
+    var userName = req.body.quizName
+    if (userName == '') {
+        userName = 'unknown user'
+    }
+
+    console.log(quizName) // TEMP
+    
+    var firstQuestion = await db.get(`${quizName}-question-1`)
+    if (firstQuestion == null) {
+        firstQuestion = "Sorry, this thing your on doesn't exist :("
+    }
+    await db.hGet('questionCount', quizName, '0') 
+    
+    if (quizName == inCase) {
+        res.send(`<!DOCTYPE html>
+        <html>
+        <head>
+            <title>Error</title>
+        </head>
+        <body>
+            <h1>${inCase}</h1>
+                <strong>Don't worry; click <em><a href="http://localhost">here</a></em> to return to the homepage.</strong>
+        </body>
+       </html>`)
+    } else {
+
+        res.send(`<!DOCTYPE html>
+        <html>
+        <head>
+            <title>Question 1 in ${quizName}</title>
+        </head>
+        <body>
+            <h1>Question 1 in ${quizName}</h1>
+                <strong>${firstQuestion}</strong>
+                
+                <form action="/take/${quizName}/${userName}/2" method="post">
+                    <h2>Type the Answer Below:</h2>
+                    <input type="text" id="userInput" name="userInput"><br />
+                    <br />
+                    <button type="submit">Click Here to Submit Your Answer</button>
+                </form>
+        </body>
+        </html>`)
+
+    }
+})
+
+app.post('/take/:quizName/:userName/:questionNum', urlencodedParser, async(req, res) => {
+    var userName = req.params.userName
+    var quizName = req.params.quizName
+    var questionNum = Number(req.params.questionNum)
+    var previousQuestionNum = questionNum - 1
+
+    //These two lines save the user's input from the previous question
+    var userInput = req.body.userInput
+    await db.set(`${quizName}-${userName}-${previousQuestionNum}`, `${userInput}`)
+    
+    console.log(`The name of this quiz is: ${quizName}`) //TEMP
+    console.log(`This is question ${questionNum}`) //TEMP
+    
+    const questionText = req.body.questionText
+    const answerText = req.body.answerText
+
+    const suspicious = await db.get(`${quizName}-question-${questionNum}`, questionText)
+
+    if (suspicious == null) {
+        console.log("suspicious stuff is happening")  //TEMP
+        res.redirect(`/take/${quizName}/${userName}/score`)
+    } else {
+
+        await db.get(`${quizName}-answer-${questionNum}`, answerText)
+        await db.hSet("questionCount", quizName, questionNum) 
```

Comment:
taking the quiz should read the question count, not set it, which provides how many questions the quiz has. Then compare this `questionNum` to determine if the quiz is done.

Reply:
yes HGET retrieves the value of a field in the redis hash table. The hash table is the same thing as a map, like what we saw in python - a table of key/value pairs.